ext/intl/listformatter/listformatter_class.c

```diff
@@ -0,0 +1,241 @@
+        Z_PARAM_STRING(locale, locale_len)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG(type)
+        Z_PARAM_LONG(width)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if(locale_len == 0) {
+        locale = (char *)intl_locale_get_default();
+    }
+
+    if (locale_len > INTL_MAX_LOCALE_LEN) {
+        zend_argument_value_error(1, "Locale string too long, should be no longer than %d characters", INTL_MAX_LOCALE_LEN);
+        RETURN_THROWS();
+    }
+
+    if (strlen(uloc_getISO3Language(locale)) == 0) {
+        zend_argument_value_error(1, "\"%s\" is invalid", locale);
+        RETURN_THROWS();
+    }
+
+    UErrorCode status = U_ZERO_ERROR;
+    #if U_ICU_VERSION_MAJOR_NUM >= 67
+        if (type != ULISTFMT_TYPE_AND && type != ULISTFMT_TYPE_OR && type != ULISTFMT_TYPE_UNITS) {
+            zend_argument_value_error(2, "must be one of IntlListFormatter::TYPE_AND, IntlListFormatter::TYPE_OR, or IntlListFormatter::TYPE_UNITS");
+            RETURN_THROWS();
+        }
+        
+        if (width != ULISTFMT_WIDTH_WIDE && width != ULISTFMT_WIDTH_SHORT && width != ULISTFMT_WIDTH_NARROW) {
+            zend_argument_value_error(3, "must be one of IntlListFormatter::WIDTH_WIDE, IntlListFormatter::WIDTH_SHORT, or IntlListFormatter::WIDTH_NARROW");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_openForType(locale, type, width, &status);
+    #else
+        if (type != ULISTFMT_TYPE_AND) {
+            zend_argument_value_error(2, "contains an unsupported type. ICU 66 and below only support IntlListFormatter::TYPE_AND");
+            RETURN_THROWS();
+        }
+
+        if (width != ULISTFMT_WIDTH_WIDE) {
+            zend_argument_value_error(3, "contains an unsupported width. ICU 66 and below only support IntlListFormatter::WIDTH_WIDE");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_open(locale, &status);
+    #endif
+
+    if (U_FAILURE(status)) {
+        intl_error_set(NULL, status, "Constructor failed", 0);
+        zend_throw_exception(IntlException_ce_ptr, "Constructor failed", 0);
+        RETURN_THROWS();
+    }
+}
+
+PHP_METHOD(IntlListFormatter, format)
+{
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+    HashTable *ht;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+        Z_PARAM_ARRAY_HT(ht)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (!LISTFORMATTER_OBJECT(obj)) {
+        zend_throw_exception(NULL, "ListFormatter not properly constructed", 0);
+        RETURN_FALSE;
+    }
+
+    uint32_t count = zend_hash_num_elements(ht);
+    if (count == 0) {
+        RETURN_EMPTY_STRING();
+    }
+
+    const UChar **items = (const UChar **)safe_emalloc(count, sizeof(const UChar *), 0);
+    int32_t *itemLengths = (int32_t *)safe_emalloc(count, sizeof(int32_t), 0);
+    uint32_t i = 0;
+    zval *val;
+
+    ZEND_HASH_FOREACH_VAL(ht, val) {
+        zend_string *str_val;
+        
+        str_val = zval_get_string(val);
```

Comment:
One optimization that you could do (optional) is using `zval_get_tmp_string` instead. Although it probably doesn't matter too much here